craft_platforms/snap/_build.py

```diff
@@ -0,0 +1,177 @@
+                base,
+                message="build-base 'devel' is not valid if base is 'bare'",
+                resolution="Ensure the build-base is a supported base.",
+                docs_url="https://snapcraft.io/docs/base-snaps",
+            )
+        if build_base is None:
+            raise _errors.NeedBuildBaseError(base)
+        try:
+            return get_distro_base_from_core_base(build_base)
+        except _errors.InvalidBaseError:
+            raise _errors.InvalidBaseError(
+                build_base,
+                resolution="Ensure the build-base is a supported base.",
+                docs_url="https://snapcraft.io/docs/base-snaps",
+                build_base=True,
+            ) from None
+    if not (matches := CORE_BASE_REGEX.match(base)):
+        raise _errors.InvalidBaseError(
+            base,
+            resolution="Ensure the base is a supported base.",
+            docs_url="https://snapcraft.io/docs/base-snaps",
+        )
+    major_release = matches.group("major")
+    extra = matches.group("extra")
+    if extra and not build_base:
+        raise _errors.NeedBuildBaseError(base)
+    if not extra and build_base not in ("devel", None):
+        raise _errors.InvalidBaseError(
+            base,
+            build_base=True,
+            message=f"base {base!r} cannot use a build-base",
+            resolution="Remove the 'build-base' key.",
+            docs_url="https://snapcraft.io/docs/base-snaps",
+        )
+    if (
+        build_base
+        and (build_base_matches := CORE_BASE_REGEX.match(build_base))
+        and build_base_matches.group("extra")
+    ):
+        raise _errors.InvalidBaseError(
+            build_base,
+            build_base=True,
+            resolution="Ensure the build-base is only the core base.",
+            docs_url="https://snapcraft.io/docs/base-snaps",
+        )
+    if build_base == "devel":
+        return _distro.DistroBase("ubuntu", "devel")
+    if base in ("core", "core16"):
+        return _distro.DistroBase("ubuntu", "16.04")
+    return _distro.DistroBase("ubuntu", f"{major_release}.04")
+
+
+@typing.overload
+def get_platforms_snap_build_plan(
+    base: None,
+    *,
+    platforms: _platforms.Platforms | None,
+    build_base: str | None = None,
+    snap_type: typing.Literal["base", "kernel"],
+) -> Sequence[_buildinfo.BuildInfo]: ...
+@typing.overload
+def get_platforms_snap_build_plan(
+    base: str,
+    *,
+    platforms: _platforms.Platforms | None,
+    build_base: str | None = None,
+    snap_type: str | None = None,
+) -> Sequence[_buildinfo.BuildInfo]: ...
+def get_platforms_snap_build_plan(
+    base: str | None,
+    *,
+    platforms: _platforms.Platforms | None,
+    build_base: str | None = None,
+    snap_type: str | None = None,
+) -> Sequence[_buildinfo.BuildInfo]:
+    """Generate the build plan for a platforms-based charm."""
+    distro_base = None
+    if base is None:
+        if snap_type in ("base", "kernel"):
```

Comment:
What happens if one of these types declares a base too? Is that verboten?

Reply:
It is an error because we don't have a definition of what it means for any of these types to have a `base`.

It may change for `type: kernel` next year but is unlikely to change for `base` or `snapd`.